wicker/core/datasets.py

```diff
@@ -207,6 +271,106 @@ def schema(self) -> DatasetSchema:
+            path_factory,
+            storage,
+            columns_to_load=columns_to_load,
+            local_cache_path_prefix=local_cache_path_prefix,
+            treat_objects_as_bytes=treat_objects_as_bytes,
+            filters=filters,
+        )
+
+
+class S3Dataset(BaseDataset):
+    """Implementation for a Map-based dataset"""
+
+    def __init__(
+        self,
+        dataset_name: str,
+        dataset_version: str,
+        dataset_partition_name: str,
+        local_cache_path_prefix: str = os.getenv("TMPDIR", "/tmp"),
+        columns_to_load: Optional[List[str]] = None,
+        storage: Optional[S3DataStorage] = None,
+        s3_path_factory: Optional[S3PathFactory] = None,
+        pa_filesystem: Optional[pafs.FileSystem] = None,
+        filelock_timeout_seconds: int = FILE_LOCK_TIMEOUT_SECONDS,
+        treat_objects_as_bytes: bool = False,
+        filters=None,
+    ):
+        """Initializes an S3Dataset
+
+        :param dataset_name: name of the dataset
+        :param dataset_version: version of the dataset
+        :param dataset_partition_name: partition name
+        :param local_cache_path_prefix: Path to local cache path, if empty don't create cache
+        :param columns_to_load: list of columns to load, defaults to None which loads all columns
+        :param storage: S3DataStorage object for pulling files from cloud storage
+        :param s3_path_factory: S3PathFactory for pulling consistent paths.
+        :param pa_filesystem: Pyarrow filesystem for reading the parquet files and tables.
+        :param filelock_timeout_seconds: number of seconds after which to timeout on waiting for downloads,
+            defaults to FILE_LOCK_TIMEOUT_SECONDS
+        :param treat_objects_as_bytes: If set, don't try to decode ObjectFields and keep them as binary data
+        :param filters: Only returns rows which match the filter. Defaults to None, i.e., returns all rows.
+        :type filters: pyarrow.compute.Expression, List[Tuple], or List[List[Tuple]], optional
+        .. seealso:: `filters in <https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html>`__ # noqa
+        """
+        s3_path_factory = s3_path_factory if s3_path_factory is not None else S3PathFactory()
+        pa_filesystem = (
+            pafs.S3FileSystem(region=get_config().aws_s3_config.region) if pa_filesystem is None else pa_filesystem
+        )
+        super().__init__(
+            dataset_name,
+            dataset_version,
+            dataset_partition_name,
+            pa_filesystem,
+            s3_path_factory,
+            storage,
+            columns_to_load=columns_to_load,
+            filelock_timeout_seconds=filelock_timeout_seconds,
+            local_cache_path_prefix=local_cache_path_prefix,
+            treat_objects_as_bytes=treat_objects_as_bytes,
+            filters=filters,
+        )
+        self._s3_path_factory = s3_path_factory
+        self.filters = filters
+
```

Comment:
The `arrow_table` function in the `S3Dataset` class has just a little bit of S3-specific logic in it, so I kept its override here. However, the `__len__` and `__getitem__` functions can be inherited perfectly from `BaseDataset`.